typescript/src/schema/api.ts

```diff
@@ -83,3 +91,48 @@ export class API {
     );
   }
 }
+
+function getResourceMap(
+  fd: plugin.google.protobuf.IFileDescriptorProto[]
+): ResourceMap {
+  const resourceMap: ResourceMap = {};
+  for (const fileDescriptor of fd) {
+    const messages = fileDescriptor
+      .messageType!.filter(message => message.name)
+      .reduce(
+        (map, message) => {
+          map['.' + fileDescriptor.package! + '.' + message.name!] = message;
+          return map;
+        },
+        {} as MessagesMap
+      );
+    for (const property of Object.keys(messages)) {
+      const m = messages[property];
+      if (m && m.options) {
+        const option = m.options;
+        if (option && option['.google.api.resource']) {
+          const opt = option['.google.api.resource'];
+          const oneResource = option[
+            '.google.api.resource'
+          ] as ResourceDescriptor;
+          if (opt.type) {
+            const arr = opt.type.match(/\/([^.]+)$/);
+            if (arr) {
+              oneResource.name = arr[arr.length - 1];
+            }
+          }
+          const pattern = opt.pattern;
+          if (pattern && pattern[0]) {
+            const params = pattern[0].match(/{[a-zA-Z]+}/g) || [];
+            for (let i = 0; i < params.length; i++) {
+              params[i] = params[i].replace('{', '').replace('}', '');
+            }
+            oneResource.params = params;
+          }
+          resourceMap[opt.type!] = oneResource;
```

Comment:
What if `oneResource` still has name or params undefined? We probably should warn and should not save it.

Reply:
very helpful! Added logics to check name and params